docs/en/3-Pipelines/Kubeflow-Pipelines.md

```diff
@@ -230,20 +247,478 @@ run = client.run_pipeline(
+    ![A Kubeflow Pipeline](../images/kubeflow_pipelines_notebook_server_image.png)
+
+### Defining components directly in YAML
+
+Components can be defined directly with a YAML file, where the designer can run
+terminal commands from a given Docker image. This can be a great way to make
+components that use non-python code from existing containers. Like with all
+components, we can pass both arguments and data files into/out of the component.
+For example:
+
+```yaml
+name: Concat Strings
+inputs:
+  - {
+      name: Input text 1,
+      type: String,
+      description: "Some text to echo into a terminal and tee to a file",
+    }
+  - {
+      name: Input text 2,
+      type: String,
+      description: "Some text to echo into a terminal and tee to a file",
+    }
+outputs:
+  - { name: Output filename, type: String }
+implementation:
+  container:
+    image: bash:5
+    command:
+      - bash
+      - -ex
+      - -c
+      - |
+        echo "$0 | $1" | tee $2
+      - { inputValue: Input text 1 }
+      - { inputValue: Input text 2 }
+      - { outputPath: Output filename }
+```
+
+concatenates two stings similar to our lightweight example above. We then define
+a component in python from this YAML by:
+
+```python
+from kfp.components import load_component_from_file
+echo_and_tee = load_component_from_file('path/to/echo_and_tee.yaml')
+
+@dsl.pipeline
+def my_pipeline():
+    echo_and_tee_task_1 = echo_and_tee("My text to echo")
+
+    # A second use that consumes the return of the first one
+    echo_and_tee_task_2 = echo_and_tee(echo_and_tee_task_1.output)
+```
+
+See
+[this example](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb)
+for more details on using existing components.
+
+### Reusing existing components
+
+Similar to well abstracted functions, well abstracted components can reduce the
+amount of code you have to do on any given project. For example, rather than
+teaching your machine learning `train_model` component to then save the model to
+MinIO, you can instead have `train_model` return the model and then Kubeflow
+Pipelines pass the model to a reusable `copy_to_minio` component. This reuse
+pattern applies to components defined through any means (containers,
+lightweight, or YAML). An example of this is given
+[here](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb),
+which reuses provided components for simple file IO tasks.
```

Comment:
```suggestion
Similar to well abstracted functions, well abstracted components can reduce the
amount of code you have to write for any given project. For example, rather than
teaching your machine learning `train_model` component to also save the resulting model to
MinIO, you can instead have `train_model` return the model and then Kubeflow
Pipelines can pass the model to a reusable `copy_to_minio` component. This reuse
pattern applies to components defined through any means (containers,
lightweight, or YAML). [Take a look at our example notebook](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb),
which reuses provided components for simple file IO tasks.
```